### PR TITLE
feat: user executor

### LINKED ISSUE #320: feat: include a test from another file into a test suite
**Is your feature request related to a problem? Please describe.**
When you want to re-use a test in several test suites, you have two possibilites:
- duplicate your test in each test suite
- create a specific test suite for this test and create a symlink to this test suite

**Describe the solution you'd like**
Be able to include a test inside a test suite at a specific location

**Additional context**
At the moment, I used a `common` directory where I put all my common tests in dedicated test suites. Then I use a symlink in each test directory where I want to use this test.

### PR DESCRIPTION
**Pull request by yesnault**
close #320

Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

File: process_teststep.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (v *Venom) RunTestStep(ctx context.Context, e ExecutorRunner, ts *TestSuite
 			if oDir == "" {
 				oDir = "."
 			}
-			filename := path.Join(oDir, fmt.Sprintf("%s.%s.step.%d.dump.json", slug.Make(ts.ShortName), slug.Make(tc.Name), stepNumber))
+			filename := path.Join(oDir, fmt.Sprintf("%s.%s.step.%d.dump.json", StringVarFromCtx(ctx, "venom.testsuite.shortName"), slug.Make(tc.Name), stepNumber))
```
Comment (fsamin) [inline process_teststep.go:68]: where is the slug.Make ?

File: types_executor.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	return d
+}
+
+type UserExecutor struct {
+	Executor     string             `json:"executor" yaml:"executor"`
+	Input        H                  `json:"input" yaml:"input"`
+	RawTestSteps []json.RawMessage  `json:"steps" yaml:"steps"`
+	Output       UserExecutorOutput `json:"output" yaml:"output"`
+	v            *Venom             `json:"-" yaml:"-"`
+}
```
Comment (fsamin) [inline types_executor.go:135]: missing output

File: types_executor.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+				vrs.Add("input."+k, vl)
+			} else {
+				// default value from executor
+				vrs.Add("input."+k, v)
+			}
+		} else {
+			vrs.Add(k, v)
+		}
+	}
+
```
Comment (fsamin) [inline types_executor.go:160]: don't we need to interpolate inputs ?

File: types_executor.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	Debug(ctx, "running user executor %v\n", tc.Name)
+	Debug(ctx, "with vars: %v", vrs)
+
+	ux.v.runTestSteps(ctx, tc)
+	result := tc.computedVars
+	if len(tc.Errors) > 0 || len(tc.Failures) > 0 {
+		return result, fmt.Errorf("failed")
+	}
+	return result, nil
```
Comment (fsamin) [inline types_executor.go:180]: there is way to much data in computedVars, we have to let the user format his own output

File: venom.go [hunk 5]
```diff
@@ -119,12 +124,48 @@ func (v *Venom) GetExecutorRunner(ctx context.Context, t TestStep, h H) (context
...
+
+		v.RegisterExecutor(name, ux)
+	}
+	return nil
+}
+
+func (v *Venom) registerPlugin(ctx context.Context, name string, vars map[string]string) error {
+	workdir := vars["venom.testsuite.workdir"]
+	// try to load from testsuite path
+	p, err := plugin.Open(workdir + "/executors/" + name + ".so")
```
Comment (fsamin) [inline venom.go:160]: use filepath.Join

File: venom.go [hunk 6]
```diff
@@ -119,12 +124,48 @@ func (v *Venom) GetExecutorRunner(ctx context.Context, t TestStep, h H) (context
...
+	if err != nil {
+		// try to load from venom binary path
+		p, err = plugin.Open("/executors/" + name + ".so")
```
Comment (fsamin) [inline venom.go:163]: use filepath.Join

File: types_executor.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	return d
+}
+
+type UserExecutor struct {
+	Executor     string             `json:"executor" yaml:"executor"`
+	Input        H                  `json:"input" yaml:"input"`
+	RawTestSteps []json.RawMessage  `json:"steps" yaml:"steps"`
+	Output       UserExecutorOutput `json:"output" yaml:"output"`
+	v            *Venom             `json:"-" yaml:"-"`
```
Comment (fsamin) [inline types_executor.go:134]: UserExecutor don't have to have venom itself as attribute, it is so weird

File: types_executor.go [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+	tc.originalName = tc.Name
+	tc.Name = slug.Make(tc.Name)
+	tc.Vars.Add("venom.testcase", tc.Name)
+	tc.computedVars = H{}
+
+	Debug(ctx, "running user executor %v\n", tc.Name)
+	Debug(ctx, "with vars: %v", vrs)
+
+	ux.v.runTestSteps(ctx, tc)
```
Comment (fsamin) [inline types_executor.go:175]: humm, it's really not the role to an executor to call venom behavior. I think you should patch the venom step executor to deal with user define executor but not the opposite

File: venom.go [hunk 9]
```diff
@@ -119,12 +124,48 @@ func (v *Venom) GetExecutorRunner(ctx context.Context, t TestStep, h H) (context
...
 
+	for _, f := range executorsPath {
+		log.Info("Reading ", f)
+		btes, err := ioutil.ReadFile(f)
+		if err != nil {
+			return errors.Wrapf(err, "unable to read file %q", f)
+		}
+
+		ux := UserExecutor{}
+		if err := yaml.Unmarshal([]byte(btes), &ux); err != nil {
```
Comment (fsamin) [inline venom.go:142]:
```suggestion
		if err := yaml.Unmarshal(btes, &ux); err != nil {
```

File: venom.go [hunk 10]
```diff
@@ -119,12 +124,48 @@ func (v *Venom) GetExecutorRunner(ctx context.Context, t TestStep, h H) (context
...
+			Error(context.Background(), "file %q content: %s", f, btes)
```
Comment (fsamin) [inline venom.go:143]:
```suggestion
```

File: venom.go [hunk 11]
```diff
@@ -119,12 +124,48 @@ func (v *Venom) GetExecutorRunner(ctx context.Context, t TestStep, h H) (context
...
+			return errors.Wrapf(err, "error while unmarshal file %q", f)
```
Comment (fsamin) [inline venom.go:144]:
```suggestion
			return fmt.Errorf(err, "unable to parse file %q: %v", f, err)
```

**Review by fsamin — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -126,6 +127,45 @@ testcases:
...
+file `executors/customA.yml`
+```yml
+executor: customA
+input:
+  myarg: {}
+steps:
+- script: echo "foo:{{.input.myarg}}"
+  assertions:
+  - result.code ShouldEqual 0
+```
```
[the remark this answers is not in the document]
Comment (yesnault) [reply README.md:145]: fixed

File: types_executor.go (hunk 3 above)
Comment (yesnault) [reply types_executor.go:174]: already done

**Review by fsamin — APPROVED** [review]

**Pull request merged**